### PR TITLE
collapse highlevel and lowlevel Simulator classes

### LINKED ISSUE #1053: Factor out high-level Simulator class
The Simulator class in [ancestry.py](https://github.com/tskit-dev/msprime/blob/6718266c48706f964364ced485d34c399a94c0c9/msprime/ancestry.py#L556) is basically a thin wrapper around the low-level Simulator object in the Python-C interface. It doesn't really do anything much add a bunch of boilerplate, and the functionality it does have could easily be pulled out into functions.

It would be nice if we could set get rid of the high level Simulator class entirely, and change the low-level Simulator so that all the methods like ``get_num_breakpoints()`` become properties (that is, so they are basically like the high-level simulator).

It would be a bit of refactoring, but it would remove a lot of boilerplate and complexity, and it would help people understand how things fit together better.

### PR DESCRIPTION
**Pull request by grahamgower**
closes #1053

Not done yet, but these are the changes most likely to cause merge conflicts. I wasn't sure which of the remaining code should be in simulator_factory or the Simulator subclass, so figured I'd send this in for some feedback.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
YES! Thanks @grahamgower, this looks just right! Some great stuff in here.

Do you mind if we park this for a few days while a few other PRs are working their way through the pipeline? I have #1025, #970 and #1066 which will all conflict a little bit with this, so it'd be nice to flush them through first.

But this is great, I'd expect only minor changes from what you have.

**Comment by grahamgower** [comment]
No probelm @jeromekelleher, just ping me once those other PRs have merged and let me know what you'd like changed here.

**Comment by grahamgower** [comment]
Conflict resolution was a little tricky in some bits, so this wants careful checking that I didn't remove anything that was just introduced.

File: tests/test_lowlevel.py [hunk 1]
```diff
@@ -546,45 +546,46 @@ def verify_squashed_edges(self, sorted_edges):
...
-            self.assertGreaterEqual(sim.get_num_rejected_common_ancestor_events(), 0)
-        self.assertGreater(sim.get_num_avl_node_blocks(), 0)
-        self.assertGreater(sim.get_num_segment_blocks(), 0)
-        self.assertGreater(sim.get_num_node_mapping_blocks(), 0)
-
-        edges = sim.get_edges()
+        if sim.model == "hudson":
+            self.assertEqual(sim.num_rejected_common_ancestor_events, 0)
+        elif sim.model in ["smc", "smc_prime"]:
+            self.assertGreaterEqual(sim.num_rejected_common_ancestor_events, 0)
```
Comment (jeromekelleher) [inline tests/test_lowlevel.py:571]:
```suggestion
        model_name = sim.model["name"]
        if model_name == "hudson":
            self.assertEqual(sim.num_rejected_common_ancestor_events, 0)
        elif model_name in ["smc", "smc_prime"]:
            self.assertGreaterEqual(sim.num_rejected_common_ancestor_events, 0)
```

File: msprime/ancestry.py [hunk 2]
```diff
@@ -695,195 +683,171 @@ def simulate(
 # and replace with the low-level class. If we change the low-level class
```
Comment (jeromekelleher) [inline msprime/ancestry.py:683]: Can delete this comment now

File: tests/test_lowlevel.py (hunk 1 above)
Comment (grahamgower) [reply tests/test_lowlevel.py]: Ah, nice catch!

**Comment by grahamgower** [comment]
squashed and ready...

**Comment by jeromekelleher** [comment]
Thanks again Graham, this is fantastic.

**Pull request merged**